File: cw2/python/threads.py

import sys
import threading
import time

class Matrix:
    def __init__(self, rows, cols):
        self.rows = rows
        self.cols = cols
        self.data = [[0] * cols for _ in range(rows)]
        self.lock = threading.Lock() 

    def set(self, row, col, value):
        with self.lock:
            self.data[row][col] = value

    def get(self, row, col):
        return self.data[row][col]

    def display(self):
        for row in self.data:
            print(row)


def read_matrix (filename):
    with open(filename, "r") as file:
        rows = int(file.readline())
        cols = int(file.readline())
        matrix = Matrix(rows, cols)
        for i in range(rows):
            line = file.readline().split()
            for j in range(cols):
                matrix.set(i, j, float(line[j]))
        return matrix 


def create_intervals(num_threads, max_index):
    intervals = []

    rest = max_index % num_threads
    part = max_index // num_threads
    addition = 0

    for i in range(num_threads):
        if rest > 0:
            addition += 1
            rest -= 1
            intervals.append((i * part + addition, i * part + part + addition))
        else:
            intervals.append((i * part + addition, i * part + part + addition))

    print(intervals)
    return intervals



def calculate_result(A, B, C, intervals, threadId):

    start_index, end_index = intervals

    for i in range(start_index, end_index):
        row = i // C.cols
        col = i % C.cols
        value = 0
        for j in range(A.cols):
            value += A.get(row, j) * B.get(j, col)
        C.set(row, col, value)


def calculate_forbenius_norm(C, intervals, threadId, result):

    result[threadId] = 0
    start_index, end_index = intervals

    for i in range(start_index, end_index):
        row = i // C.cols
        col = i % C.cols
        result[threadId] += C.get(row, col) ** 2


def main():

    if len(sys.argv) != 3:
        print("Usage: threads.py <number of threads> <mode>")
        sys.exit(1)
    
    num_threads = int(sys.argv[1])
    mode = sys.argv[2]

    filenameA = ""
    filenameB = ""

    if mode == "base":
        filenameA = "data/A.txt"
        filenameB = "data/B.txt"
    elif mode == "gen":
        filenameA = "data/genA.txt"
        filenameB = "data/genB.txt"
    else:
        print("Invalid mode")
        sys.exit(1)

    A = read_matrix(filenameA)
    B = read_matrix(filenameB)

    if A.cols != B.rows:
        print("Invalid matrix dimensions")
        sys.exit(1)
    
    C = Matrix(A.rows, B.cols)

    intervals = create_intervals(num_threads, C.rows * C.cols)

    begin = time.time()
    threads = []

    for i in range(num_threads):
        t = threading.Thread(target=calculate_result, args=(A, B, C, intervals[i], i))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()
    
    # C.display()

    forbenius = [0] * num_threads

    for i in range(num_threads):
        t = threading.Thread(target=calculate_forbenius_norm, args=(C, intervals[i], i, forbenius))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    forbenius_norm = 0

    for i in range(num_threads):
        forbenius_norm += forbenius[i]
    
    forbenius_norm = forbenius_norm ** 0.5

    end = time.time()

    print("Time: ", end - begin)

    print("Forbenius norm: ", forbenius_norm)

if __name__ == "__main__":
    main()

File: cw2/java/Threads.java
import java.io.File;
import java.io.IOException;
import java.util.Scanner;
import java.util.concurrent.atomic.DoubleAdder; 

public class Threads {

    public static void main(String[] args) {

        if(args.length != 2) {
            System.out.println("Not enough arguments");
            System.out.println("Usage: java Threads <number of threads> <number of iterations> <mode>");
            System.out.println("Mode: base, gen");
            System.exit(1);
        }

        int numThreads = Integer.parseInt(args[0]);
        String mode = args[1];

        String filenameA;
        String filenameB;

        if(mode.equals("base")) {
            filenameA = "data/A.txt";
            filenameB = "data/B.txt";
        } else if(mode.equals("gen")) {
            filenameA = "data/genA.txt";
            filenameB = "data/genB.txt";
        } else {
            System.out.println("Invalid mode");
            return;
        }

        Matrix A = null;
        Matrix B = null;
        Matrix C;

        try (Scanner scanner = new Scanner(new File(filenameA))) {
         
            int rows = scanner.nextInt();
            int cols = scanner.nextInt();
            A = new Matrix(rows, cols);

            for(int i = 0; i < rows; i++) {
                for(int j = 0; j < cols; j++) {
                    A.set(i, j, scanner.nextDouble());
                }
            }

        } catch (IOException e) {
            System.err.println("Error reading the file: " + e.getMessage());
        }


        try (Scanner scanner = new Scanner(new File(filenameB))) {
         
            int rows = scanner.nextInt();
            int cols = scanner.nextInt();
            B = new Matrix(rows, cols);

            for(int i = 0; i < rows; i++) {
                for(int j = 0; j < cols; j++) {
                    B.set(i, j, scanner.nextDouble());
                }
            }
            
        } catch (IOException e) {
            System.err.println("Error reading the file: " + e.getMessage());
        }

        C = new Matrix(A.getRows(), B.getCols());

        int[] intervals = calculateIntervals(numThreads, C.getRows() * C.getCols());

        Thread[] threads = new Thread[numThreads];

        long startTime = System.currentTimeMillis();

        for(int i = 0; i < numThreads; i++) {
            threads[i] = new Worker(A, B, C, intervals, i);
            threads[i].start();
        }

        for(int i = 0; i < numThreads; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                System.err.println("Error joining the threads: " + e.getMessage());
            }
        }

        // C.print();

        DoubleAdder sum = new DoubleAdder();

        for(int i = 0; i < numThreads; i++) {
            threads[i] = new ForbeniusWorker(C, intervals, sum, i);
            threads[i].start();
        }

        for(int i = 0; i < numThreads; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                System.err.println("Error joining the threads: " + e.getMessage());
            }
        }

        long endTime = System.currentTimeMillis();

        System.out.println("Forbenius norm: " + Math.sqrt(sum.doubleValue()));
        System.out.println("Time: " + (endTime - startTime) + "ms");
    }

    private static int[] calculateIntervals(int numOfThreads, int maxIndex) {

        int[] intervals = new int[numOfThreads * 2];

        int rest = maxIndex % numOfThreads;
        int part = (maxIndex - rest) / numOfThreads;
        int addition = 0;
        int index = 0;

        for(int i = 0; i < numOfThreads; i++) {
            if(rest > 0) {
                intervals[index] = i * part + addition;    
                intervals[++index] = i * part + part + addition + 1;
                addition++;
                rest--;
            } else {
                intervals[index] = i * part + addition;
                intervals[++index] = i * part + part + addition;
            }

            index++;
        }

        // for(int i = 0; i < intervals.length; i++) {
        //     System.out.println(intervals[i] + " - " + intervals[++i]);
        // }

        return intervals;
    }

    private static class Worker extends Thread {
        private Matrix A;
        private Matrix B;
        private Matrix C;
        private int[] intervals;
        private int threadId;

        public Worker(Matrix A, Matrix B, Matrix C, int[] intervals, int threadId) {
            this.A = A;
            this.B = B;
            this.C = C;
            this.intervals = intervals;
            this.threadId = threadId;
        }

        public void run() {
            for(int i = intervals[threadId * 2]; i < intervals[threadId * 2 + 1]; i++) {
                int row = i / C.getCols();
                int col = i % C.getCols();
                double sum = 0;

                for(int j = 0; j < A.getCols(); j++) {
                    sum += A.get(row, j) * B.get(j, col);
                }
                // System.out.println("Thread " + Thread.currentThread().getId() + " is working on row " + row + " and col " + col + " with sum " + sum);
                C.set(row, col, sum);
            }
        }
    }

    private static class ForbeniusWorker extends Thread{

        private Matrix C;
        private int[] intervals;
        private DoubleAdder sum;
        private int threadId;

        public ForbeniusWorker(Matrix C, int[] intervals, DoubleAdder sum, int threadId) {
            this.C = C;
            this.intervals = intervals;
            this.sum = sum;
            this.threadId = threadId;
        }

        public void run() {
            for(int i = intervals[threadId * 2]; i < intervals[threadId * 2 + 1]; i++) {
                int row = i / C.getCols();
                int col = i % C.getCols();
                sum.add(Math.abs(C.get(row, col) * C.get(row, col)));
            }
        }
    }
}

class Matrix {
    private double[][] matrix;
    private int rows;
    private int cols;

    public Matrix(int rows, int cols) {
        matrix = new double[rows][cols];
        this.rows = rows;
        this.cols = cols;
    }

    synchronized public void set(int row, int col, double value) {
        matrix[row][col] = value;
    }

    public double get(int row, int col) {
        return matrix[row][col];
    }

    public int getRows() {
        return rows;
    }

    public int getCols() {
        return cols;
    }

    public void print() {
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }
    }
}